hudi-common/src/main/java/org/apache/hudi/common/model/DefaultHoodieRecordPayload.java

```diff
@@ -86,30 +86,26 @@ public Option<IndexedRecord> getInsertValue(Schema schema, Properties properties
     GenericRecord incomingRecord = HoodieAvroUtils.bytesToAvro(recordBytes, schema);
     eventTime = updateEventTime(incomingRecord, properties);
 
-    return isDeleteRecord(incomingRecord, properties) ? Option.empty() : Option.of(incomingRecord);
+    return isDeleted(schema, properties) ? Option.empty() : Option.of(incomingRecord);
   }
 
-  /**
-   * @param genericRecord instance of {@link GenericRecord} of interest.
-   * @param properties payload related properties
-   * @returns {@code true} if record represents a delete record. {@code false} otherwise.
-   */
-  protected boolean isDeleteRecord(GenericRecord genericRecord, Properties properties) {
-    final String deleteKey = properties.getProperty(DELETE_KEY);
+  @Override
+  protected boolean isDeleteRecord(GenericRecord record, Properties props) {
+    final String deleteKey = props.getProperty(DELETE_KEY);
     if (StringUtils.isNullOrEmpty(deleteKey)) {
-      return isDeleteRecord(genericRecord);
+      return super.isDeleteRecord(record, props);
```

Comment:
Is this line the actualy fix, I didn't see the props got used by the super method, so do we still need to pass around all the props here?

Reply:
The implementation of this method is kept the same.  However, before my fix, this method `isDeleteRecord(GenericRecord genericRecord, Properties properties)` was a protected method not overriding any method in super class, and it was not called in the constructor to determine whether a record is a delete.

After the fix, the method `isDeleteRecord(GenericRecord genericRecord, Properties properties)` overrides the same in the super class, and is now called in the constructor (see super class `BaseAvroPayload` constructor, where the following is called)
```
this.isDeletedRecord = record == null || isDeleteRecord(record, props);
```

Reply:
But the `BaseAvroPayload.isDeleteRecord` does not really use the passed in props, so do we still to change all the constructors just for passing the props, it seems just an additional invocation of `BaseAvroPayload.isDeleteRecord` could fix the problem and this method does not need any props actually.

We can address the props issue is separate task maybe.

Reply:
And these 2 options are not even included in any configuration class, I don't know why they are located in the
`DefaultHoodieRecordPayload`, if the logic is customized just for `DefaultHoodieRecordPayload`, we should not invoke it in the parent constructor because not all payloads are sub-class of `DefaultHoodieRecordPayload`

- `hoodie.payload.delete.field`
- `hoodie.payload.delete.marker`

Reply:
The `BaseAvroPayload.isDeleted(Schema schema, Properties props)` already passes around the properties, so can we just overwrite it for `DefaultHoodieRecordPayload` and there is no need to change all the constructors?

Reply:
`BaseAvroPayload.isDeleteRecord(GenericRecord genericRecord, Properties props)` uses the props, and it is called by `this.isDeletedRecord = record == null || isDeleteRecord(record, props);` in the constructor.  `BaseAvroPayload.isDeleteRecord(GenericRecord genericRecord)` is only for backwards compatibility, called from user-implemented record payload implementation.
<img width="1403" alt="Screenshot 2023-10-25 at 01 20 47" src="https://github.com/apache/hudi/assets/2497195/be90adfb-2df4-4d51-973f-b2a07665b769">

Reply:
The intention is that `isDeleted(Schema schema, Properties props)` should not deserialize the bytes for perf reason so we cannot implement the custom delete marker there for `DefaultHoodieRecordPayload`.  The way now in this fix is to check the delete key and marker in the constructor before the Avro record is serialized; that’s why props needs to be passed in to the constructor.